src/Interpreters/ActionsDAG.cpp

```diff
@@ -1718,6 +1718,14 @@ ActionsDAG::SplitResult ActionsDAG::split(std::unordered_set<const Node *> split
     /// List of nodes from current actions which are not inputs, but will be in second part.
     NodeRawConstPtrs new_inputs;
 
+    /// Avoid new inputs to have the same name as existing inputs.
+    /// It's allowed for DAG but may break Block invariant 'columns with identical name must have identical structure'.
```

Comment:
Why inputs can break invariant? Is it for case when inputs just bypassed and explicitly added to outputs?

Reply:
We can have a DAG like
```
INPUT x UInt32 -> cast(x as String) -> ALIAS x String
               -> sin(x)            -> ALIAS y
```

and if we want to split calculation of `x + 1` to second dag, it would be like
```
INPUT x UInt32 -> sin(x) -> ALIAS y
INPUT x String
```

This is a correct DAG, but we can't have columns with the same names and different types in Block.